runtime/near-vm/engine/src/universal/code_memory.rs

```diff
@@ -2,170 +2,254 @@
-            executable_section_result.push(s);
+    fn as_slice_mut(&mut self) -> &mut [u8] {
+        unsafe {
+            // SAFETY: We have made sure that this is the only reference to the memory region by
+            // requiring a mutable self reference.
+            std::slice::from_raw_parts_mut(self.map, self.size)
         }
+    }
 
-        self.start_of_nonexecutable_pages = bytes;
-
-        if !data_sections.is_empty() {
-            // Data sections have different page permissions from the executable
-            // code that came before it, so they need to be on different pages.
-            let padding = round_up(bytes, page_size) - bytes;
-            buf = buf.split_at_mut(padding).1;
-
-            for section in data_sections {
-                let section = &section.bytes;
-                assert_eq!(buf.as_mut_ptr() as usize % DATA_SECTION_ALIGNMENT, 0);
-                let len = round_up(section.len(), DATA_SECTION_ALIGNMENT);
-                let (s, next_buf) = buf.split_at_mut(len);
-                buf = next_buf;
-                s[..section.len()].copy_from_slice(*section);
-                data_section_result.push(s);
+    /// Ensure this CodeMemory is at least of the requested size.
+    ///
+    /// This may invalidate any data previously written into the mapping.
+    pub fn resize(mut self, size: usize) -> rustix::io::Result<Self> {
+        if self.size < size {
+            let source_pool = self.source_pool.take();
+            // Ideally we would use mremap, but see
+            // https://bugzilla.kernel.org/show_bug.cgi?id=8691
+            unsafe {
+                mm::munmap(self.map.cast(), self.size)?;
+                std::mem::forget(self);
             }
+            Self::create(size).map(|mut m| {
+                m.source_pool = source_pool;
+                m
+            })
+        } else {
+            Ok(self)
         }
+    }
 
-        Ok((function_result, executable_section_result, data_section_result))
+    /// Write to this code memory from the beginning of the mapping.
+    ///
+    /// # Safety
+    ///
+    /// At the time this method is called, there should remain no dangling readable/executable
+    /// references to this `CodeMemory`, for the original code memory that those references point
+    /// to are invalidated as soon as this method is invoked.
+    pub unsafe fn writer(&mut self) -> CodeMemoryWriter<'_> {
+        self.executable_size = 0;
+        CodeMemoryWriter { memory: self, offset: 0 }
     }
 
-    /// Apply the page permissions.
-    pub fn publish(&mut self) {
-        if self.mmap.is_empty() || self.start_of_nonexecutable_pages == 0 {
-            return;
+    /// Publish the specified number of bytes as executable code.
+    ///
+    /// # Safety
+    ///
+    /// Calling this requires that no mutable references to the code memory remain.
+    pub unsafe fn publish(&mut self) -> Result<(), CompileError> {
```

Comment:
I feel like there are other kinds of invariants here that must be maintained (now that I think about this with a fresh head). For example, the executable code must be valid in various ways as well. One might argue that this would mean that `write_executable` needs to be `unsafe`, but that isn’t strictly a dangerous operation if `publish` isn't ever called for whatever reason.

Reply:
Good point! I hadn’t thought of that reason for unsafety :) (though OTOH `writer()` could probably be safe, for the same reason, as running code that had been publish’d actually does require a raw pointer deref)

(Technically we could also push the responsibility for executing the published code to the place where the raw pointers are dereferenced, but as you say it feels better to put that responsibility one step up to the publishing time, especially as we can hope for a better API than raw pointers when we have time to do lots of other refactorings)